### PR TITLE
firewaller handles subnet changes

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

A few key changes so that the firewaller correctly listens to subnet changes in a consuming model, and thus creates the correct ingress rules in a consuming model.

A large chunk of the work is boilerplate - the IngressSubnetsForRelation method is moved from the RemoteRelations facade to the RemoteFirewaller facade.

The remote relation handling of the remote firewaller now uses proper workers to:
- wait for remote relation to be fully registered in both models
- watch relation changes
- watch subnet changes

## QA steps

bootstrap aws
deploy mysql
switch controller
deploy mediawiki
relate mediawiki:db default.mysql
add-space default
add-subnet <172.31.0.0/20> default

The above subnet is found by looking at AWS console. After it is added, use the console to check that the security group on the mysql machine has the subnet added to its ingress rules.

### CODE REVIEW

File: worker/firewaller/firewaller.go [hunk 1]
```diff
@@ -1100,61 +1126,238 @@ func diffRanges(currentRules, wantedRules []network.IngressRule) (toOpen, toClos
...
+			// We stop the watcher here as it is tied to the remote facade
+			// which is closed as soon as we return.
+			worker.Stop(subnetsWatcher)
+			return rd.catacomb.ErrDying()
+		case <-subnetsWatcher.Changes():
+			if rd.remoteRelationId == nil {
+				// relation not ready yet.
+				continue
+			}
+			logger.Debugf("subnets changed in model %v", rd.remoteModelUUID)
```
Comment (babbageclunk) [inline worker/firewaller/firewaller.go]: Wouldn't you need to do something here? Oh, I think I understand - we go around the loop again and that calls updateNetworks. But we don't bother if there's no `remoteRelationId` yet.

File: worker/firewaller/firewaller.go [hunk 2]
```diff
@@ -1100,61 +1126,238 @@ func diffRanges(currentRules, wantedRules []network.IngressRule) (toOpen, toClos
...
+		fw:              fw,
+		tag:             tag,
+		relationReady:   relationReady,
+		remoteModelUUID: rel.SourceModelUUID,
+	}
+
+	err := catacomb.Invoke(catacomb.Plan{
+		Site: &poller.catacomb,
+		Work: func() error {
+			return poller.pollLoop()
```
Comment (babbageclunk) [inline worker/firewaller/firewaller.go]: Same here - can this just be `Work: poller.pollLoop,`?

File: worker/firewaller/firewaller.go [hunk 3]
```diff
@@ -116,28 +120,38 @@ type Firewaller struct {
...
-		unitds:                make(map[names.UnitTag]*unitData),
-		applicationids:        make(map[names.ApplicationTag]*applicationData),
-		applicationIngress:    make(map[names.RelationTag]*remoteRelationData),
-		remoteRelationsChange: make(chan *remoteRelationData, 1),
-		exposedChange:         make(chan *exposedChange),
+		firewallerApi:              cfg.FirewallerAPI,
+		remoteRelationsApi:         cfg.RemoteRelationsApi,
+		environFirewaller:          cfg.EnvironFirewaller,
+		environInstances:           cfg.EnvironInstances,
+		newRemoteFirewallerAPIFunc: cfg.NewRemoteFirewallerAPIFunc,
```
Comment (wallyworld) [inline worker/firewaller/firewaller.go]: correct!

File: apiserver/remotefirewaller/state.go [hunk 4]
```diff
@@ -4,21 +4,63 @@
...
+
+func (st stateShim) GetRemoteEntity(model names.ModelTag, token string) (names.Tag, error) {
+	r := st.State.RemoteEntities()
+	return r.GetRemoteEntity(model, token)
+}
+
+func (st stateShim) AllSubnets() (subnets []Subnet, err error) {
+	stateSubnets, err := st.State.AllSubnets()
+	if err != nil {
+		return nil, err
```
Comment (wallyworld) [inline apiserver/remotefirewaller/state.go]: fixed

File: worker/firewaller/firewaller.go [hunk 5]
```diff
@@ -1100,61 +1126,238 @@ func diffRanges(currentRules, wantedRules []network.IngressRule) (toOpen, toClos
...
 }
 
+// relationLifeChanged manages the workers to process ingress changes for
+// the specified relation.
+func (fw *Firewaller) relationLifeChanged(tag names.RelationTag) error {
+	results, err := fw.remoteRelationsApi.Relations([]string{tag.Id()})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	relErr := results[0].Error
```
Comment (wallyworld) [inline worker/firewaller/firewaller.go]: Yeah, but sadly here and elsewhere the facade returns a slice of params results, each result struct containing a result and a params.Error. To fix it, we need to stop leaking params structs out of the facade layer.

File: worker/firewaller/firewaller.go [hunk 6]
```diff
@@ -1100,61 +1126,238 @@ func diffRanges(currentRules, wantedRules []network.IngressRule) (toOpen, toClos
...
+		relationReady:       make(chan params.RemoteEntityId),
 	}
-	dead := !found || rel.Life() == params.Dead
+	fw.relationIngress[tag] = data
 
-	data, known := fw.applicationIngress[tag]
-	logger.Debugf("relation %v has existing data: %+v", tag, data)
+	err := catacomb.Invoke(catacomb.Plan{
+		Site: &data.catacomb,
+		Work: func() error {
```
Comment (wallyworld) [inline worker/firewaller/firewaller.go]: fixed

File: worker/firewaller/firewaller.go (hunk 2 above)
Comment (wallyworld) [inline worker/firewaller/firewaller.go]: fixed

File: worker/firewaller/firewaller.go (hunk 1 above)
Comment (wallyworld) [inline worker/firewaller/firewaller.go]: Correct. If subnets change or relation becomes ready, we drop out of the select, update the ingress rules, and go around again.

File: api/remotefirewaller/remotefirewaller.go [hunk 7]
```diff
@@ -26,6 +26,11 @@ func NewClient(caller base.APICallCloser) *Client {
 	return &Client{ClientFacade: frontend, facade: backend}
 }
 
+// Close implements io.Close
+func (c *Client) Close() error {
```
Comment (wallyworld) [inline api/remotefirewaller/remotefirewaller.go]: Nice pickup, there were other places where this had been done; I've fix them all

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**